tests/partitions_test.py

```diff
@@ -148,6 +150,56 @@ def test_isolate_and_reconnect_primary(network, args, **kwargs):
         assert status == TxStatus.Invalid, r
 
 
+@reqs.description("New joiner helps liveness")
+def test_new_joiner_helps_liveness(network, args):
+    primary, backups = network.find_nodes()
+
+    # Issue some transactions, so there is a ledger history that a new node must receive
+    network.txs.issue(network, number_txs=10)
+
+    # Remove a node, leaving the network frail
+    network.retire_node(primary, backups[-1])
+    backups[-1].stop()
+
+    primary, backups = network.find_nodes()
+
+    # Add a new node, but partition them before trusting them
+    new_node = network.create_node("local://localhost")
+    network.join_node(new_node, args.package, args, from_snapshot=False)
+    new_joiner_partition = [new_node]
+    new_joiner_rules = network.partitioner.partition(
+        [primary, *backups], new_joiner_partition
+    )
+
+    # Trust the new node, and wait for commit of this (but don't ask the new node itself, which doesn't know this yet)
+    network.trust_node(new_node, args, no_wait=True)
+    check_can_progress(primary)
+
+    # Partition the primary, temporarily creating a minority service that cannot make progress
+    minority_partition = backups[len(backups) // 2 :] + new_joiner_partition
+    minority_rules = network.partitioner.partition(minority_partition)
+    time.sleep(
+        network.observed_election_duration
+    )  # Ensure this node has become primary
+    with backups[0].client("user0") as c:
+        r = c.post("/app/log/private", {"id": 42, "msg": "Hello world"})
+        assert r.status_code == http.HTTPStatus.SERVICE_UNAVAILABLE
```

Comment:
Would `wait_for_new_primary()` wrapped in a `try: ... except TimeoutError:` be more consistent with that we do here?

Reply:
I'd tried something similar, but most of the existing infra helpers don't work here because they don't expect the node to be in this state. Specifically if we do anything before the timeout above, the request may timeout as its forwarded to an unreachable primary, and anything after returns this unusual error code. So I think it's nicer to be explicit (you can talk to this node, but it returns this error), rather than catching the exception thrown by something _else_ trying to talk to this node.